File: test/page/sample_generation/sample_generation.js
wn.pages['sample-generation'].onload = function(wrapper) { 
	wn.ui.make_app_page({
		parent: wrapper,
		title: 'Sample Generation',
		single_column: true
	});
	$(wrapper).find(".layout-main").html("<div class='user-settings'></div>\
	<table class='table table-bordered' style='background-color: #f9f9f9;'>\
	<tr><td>\
	<h4><i class='icon-question-sign'></i> "+wn._("Quick Help for Sample Entry Sorting and Sample Id Creation")+":</h4>\
	<ol>\
	<li>"+wn._("Data will be as per the filters you have selected")+"</li>\
	<li>"+wn._("To sort data click on downarrow icon")+"</li>\
	<li>"+wn._("For Getting a view click on eye icon")+"</li>\
	</ol>\
	</tr></td>\
	</table>");
	wrapper.user_properties = new wn.UserProperties(wrapper);			
}

wn.UserProperties = Class.extend({
	init: function(wrapper) {
		this.wrapper = wrapper;
		this.body = $(this.wrapper).find(".user-settings");
		this.make();
	},
	make: function() {
		var me = this;
		return wn.call({
			module:"test",
			page:"sample_generation",
			method: "get_tasnformer",
			callback: function(r) {
				me.options = r.message;
				me.from_date =
					me.wrapper.appframe.add_date("From Date", 
						'')
						.css("width", "200px")
						.change(function() {
							me.set_route();
						});
				me.transfromer =
					me.wrapper.appframe.add_select("Sample Id", 
						["Select Transformer..."].concat(r.message.transfromer_id))
						.css("width", "200px")
						.change(function() {
							me.set_route();
						});
				me.button=
					me.wrapper.appframe.add_button(wn._("Sort By"), function() {
							me.make_button()
						}, "icon-arrow-down").css("width", "50px")
				me.button1=
					me.wrapper.appframe.add_button(wn._("Display Results"), function() {
							me.get_result()
						}, "icon-eye-open").css("width", "50px")
			}
		});
	},
	make_button: function(){
		var me = this;
		var d = new wn.ui.Dialog({
			title: "Add New Property",
			fields: [
				{fieldtype:"Select", label:wn._("Sort By"),
					options:"\nFunctional Location\nPlant\nSub SStation", reqd:1, fieldname:"parent"},
				{fieldtype:"Select", label:wn._("Order"),
					options:"\nAscending\nDescending", reqd:1, fieldname:"parent1"},
				{fieldtype:"Select", label:wn._("Sort By"),
					options:"\nFunctional Location\nPlant\nSub SStation", reqd:1, fieldname:"parent2"},
				{fieldtype:"Select", label:wn._("Order"),
					options:"\nAscending\nDescending", fieldname:"parent3"},
				{fieldtype:"Button", label: wn._("Add"), fieldname:"add"},
			]
		});		
		d.get_input("add").click(function() {
			var args = d.get_values();
			if(!args) {
				return;
			}
			wn.call({
				module: "test",
				page: "sample_generation",
				method: "save_sorting_order",
				args: args,
				callback: function(r) {
				}
			})
			d.hide();
		});
		d.show();
	},
	get_result: function(){
		me = this;
		return wn.call({
			module:"test",
			page:"sample_generation",
			method: "get_sample_entries",
			args:{'date': this.from_date.val(), 'transfromer': this.transfromer.val()},
			callback: function(r) {
				console.log(r.message.sample_entry.length)
				if(r.message.sample_entry.length <= 0){
					msgprint(wn._("Sample Id is created for all Sample Entries"))
				}
				else{
					me.options = r.message;
					me.show_sample_entry(r.message.sample_entry)
					me.button1=
						me.wrapper.appframe.add_button(wn._("Create Sample"), function() {
							me.sample_creation()
						}, "icon-cogs").css("width", "50px")
						me.wrapper.appframe.add_button(wn._('Chart of Cost Centers'), function() { 
							wn.set_route("Form","Test Allocation Interface", "Test Allocation Interface"); 
						}, 'icon-sitemap').css("width", "50px")
				}
									
			}
		});
	},
	show_sample_entry: function(sample_entry){
		this.body.empty();
		var me = this;
		columns = [[wn._("Sample Entry"), 50], [wn._("Plant"), 50], [wn._("Functional Location"),50], [wn._("Sub Station"), 50], [wn._("Rating"),50]];

		if(sample_entry[0].sample_id){
			columns.push([wn._('Sample Id'),50])
		}
		else{
			columns.push(['',50])
		}

		this.table = $("<table class='table table-bordered'>\
			<thead><tr></tr></thead>\
			<tbody></tbody>\
		</table>").appendTo(this.body);
		
		$.each(columns, 
			function(i, col) {
			$("<th>").html(col[0]).css("width", col[1]+"px")
				.appendTo(me.table.find("thead tr"));
		});

				
		$.each(sample_entry, function(i, d) {
			var row = $("<tr>").appendTo(me.table.find("tbody"));
			
			$("<td>").html('<a href="#Form/Sample Entry/'+d.name+'">'
				+d.name+'</a>').appendTo(row);
			$("<td>").html(d.functional_location).appendTo(row);
			$("<td>").html(d.plant).appendTo(row);
			$("<td>").html(d.sub_station).appendTo(row);
			$("<td>").html(d.rating).appendTo(row);
			if(d.sample_id){
				$("<td>").html(d.sample_id).appendTo(row);
			}
			else{
				me.add_delete_button(row, d);
			}
		});
	},
	add_delete_button: function(row, d) {
		var me = this;
		$("<button class='btn btn-small btn-default'><i class='icon-trash'></i></button>")
			.appendTo($("<td>").appendTo(row))
			.attr("data-name", d.name)
			.click(function() {
				me.remove_sample_entry($(this).attr("data-name"))
			});
	},
	remove_sample_entry: function(sample_entry_id) {
		var counter = 0;
		for(dict in this.options.sample_entry){
			counter = counter + 1;
			if(this.options.sample_entry[dict]['name'] == sample_entry_id){
				this.options.sample_entry.splice(counter-1,1);
				me.show_sample_entry(this.options.sample_entry);
				break;
			}
		}
	},
	sample_creation: function(){
		me = this;
		console.log(this.options)
		return wn.call({
			module:"test",
			page:"sample_generation",
			method: "sample_generation",
			args:{'sample_entries':this.options.sample_entry},
			callback: function(r) {
				me.show_sample_entry(r.message.sample_entries)					
			}
		});
	},
	set_route: function() {
		console.log(this.from_date.val())
	}
})

File: test/page/sample_generation/sample_generation.py
import webnotes

sorting_dict = {}
sort_mapper = {'Ascending':'asc', 'Descending': 'desc'}

@webnotes.whitelist()
def get_tasnformer():
	return {
	'transfromer_id' : [d[0] for d in webnotes.conn.sql("""select name from tabTransformer""")]
	}

@webnotes.whitelist()
def save_sorting_order(parent=None, parent1=None, parent2=None, parent3=None):
	global sorting_dict 
	if parent:
		sorting_dict[parent] = parent1
	if parent2:
		sorting_dict[parent2] = parent3
	webnotes.errprint(sorting_dict)

@webnotes.whitelist()
def get_sample_entries(date, transfromer):
	conditions = get_conditions(date, transfromer)
	sort_order = get_sorting_order()

	return {
		'sample_entry' : get_sample_entries_data(conditions, sort_order)
	}

def get_conditions(date, transfromer):
	condition = ''
	conditions = []

	if date:
		conditions.append("date = '%s'"%(dormat_date(date)))

	if transfromer and transfromer!= 'Select Transformer...':
		conditions.append("functional_location= '%s'"%(transfromer))

	if conditions:
		condition = 'and '.join(conditions)

	if condition:
		return 'and '+condition

def dormat_date(date):
	import datetime
	try:
		return datetime.datetime.strptime(date, '%d-%m-%Y').strftime('%Y-%m-%d')
	except e:
		webnotes.msgprint(e,raise_exception=1)

def get_sorting_order():
	sort_list = []
	if sorting_dict:
		for key in sorting_dict:
			sort_string = ''
			sort_string += '%s %s '%(key.lower().replace(' ', '_'), sort_mapper[sorting_dict[key]])
			sort_list.append(sort_string)

	if sort_list:
		return ' order by ' + ', '.join(sort_list)

def get_sample_entries_data(conditions, sort_order):
	if not conditions:
		conditions = ''

	if not sort_order:
		sort_order = ''

	return webnotes.conn.sql("""select name, ifnull(plant,'---') as plant, 
				ifnull(functional_location,'----') as functional_location, ifnull(sub_station,'---') as sub_station, ifnull(rating, '---') as rating from `tabSample Entry` 
				where ifnull(docstatus,'') != 2 and ifnull(is_sample_generated,'No') != 'Yes'
				%(conditions)s %(sort_order)s """%{"conditions": conditions, 'sort_order': sort_order}, as_dict=1, debug=1)

@webnotes.whitelist()
def sample_generation(sample_entries):
	sample_entries = eval(sample_entries)	
	for sample in sample_entries:
		generate_sample(sample)
	return{
		'sample_entries':sample_entries
	}

def generate_sample(sample):
	from webnotes.model.doc import Document
	d = Document('Sample')
	d.sample_entry = sample['name']
	d.barcode = webnotes.conn.get_value("Sample Entry", sample['name'], 'bottle_list')
	d.client_name = webnotes.conn.get_value('Sample Entry', sample['name'], 'client_name')
	d.temperature = webnotes.conn.get_value('Sample Entry', sample['name'], 'testing_temperature')
	d.save()
	update_sample_entry(sample['name'])
	update_sample(sample, d.name)

def update_sample_entry(sample_entry_name):
	webnotes.conn.sql("""update `tabSample Entry` 
							set is_sample_generated = 'Yes' 
						where name = '%s'"""%(sample_entry_name),debug=1)
	webnotes.conn.sql("commit")

def update_sample(sample, sample_id):
	sample['sample_id'] = sample_id